RFC: Private Registry Mirrors (#285)
buildpacks/rfcs: text/0000-private-mirrors.md

# How it Works
[how-it-works]: #how-it-works

Out of scope - setting up the mirror itself. This is a separate concern that is not specific to Cloud Native Buildpacks.

Once a mirror for one or more public registry has been setup, the platform operator can configure Cloud Native Buildpacks to use the mirror. This will permit `lifecycle` to use the mirrors for all images that are requested during builds using the `CNB_REGISTRY_MIRRORS` environment variable that would otherwise be requested from the public registry.

The `CNB_REGISTRY_MIRRORS` environment variable will be a list of mirror configurations. Each mirror configuration will be a key/value pair, where the key is the registry URL and the value is the mirror URL. The key/value pairs will be separated by a semicolon (`;`).

For example, if we wanted to configure a mirror for Docker Hub and Quay.io, we could set the `CNB_REGISTRY_MIRRORS` environment variable to the following:

```
docker.io=https://docker.mirror.example.com;quay.io=https://quay.mirror.example.com
```

> When `lifecycle` requests an image during any phase (`analyze`, `export`, `extend`, etc.), it will first check the `CNB_REGISTRY_MIRRORS` environment variable. If the requested image's registry is configured in the `CNB_REGISTRY_MIRRORS` environment variable, it will use the mirror URL instead of the original registry URL.

Reviewer: Could we outline how this intersects with run.toml/stack.toml and "run image mirrors"?

> When `lifecycle` requests an image during any phase (`analyze`, `export`, `extend`, etc.), it will first check the `CNB_REGISTRY_MIRRORS` environment variable. If the requested image's registry is configured in the `CNB_REGISTRY_MIRRORS` environment variable, it will use the mirror URL instead of the original registry URL.

Reviewer: ```suggestion
When `lifecycle` requests an image during any phase (`analyze`, `restore`, `export`, `rebase`), it will first check the `CNB_REGISTRY_MIRRORS` environment variable. If the requested image's registry is configured in the `CNB_REGISTRY_MIRRORS` environment variable, it will use the mirror URL instead of the original registry URL.
```

Why should we *not* do this?

Complexity. We'll have to teach `lifecycle` how to use the configured mirrors. This will add complexity to the codebase and will require additional testing.

Breaking expectations. If a platform operator were to fall behind or modify an image in the mirror, the resulting image would not match the image built by an end user against the public registry.

# Alternatives
[alternatives]: #alternatives

- What other designs have been considered?

Platforms could use oci layout to try and fetch images from a local cache before going to the public registry. This would require the platform to have a local cache of all images that may be requested during builds. This would be difficult to maintain and would require a lot of disk space.

- Why is this proposal the best?

This may not be the best, but that is why we are proposing it. We'd like to hear from the community about other options.

- What is the impact of not doing this?

Operators will have to continue to deal with the reliability of public registries.

# Prior Art
[prior-art]: #prior-art

Discuss prior art, both the good and bad.

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> - How will will teach kaniko-style extensions to use the mirrors?

Reviewer: With luck I think the lifecycle could handle all of the mirror translations. We don't pull any images when running kaniko...

# Motivation
[motivation]: #motivation

- Why should we do this?
As a platform operator, we'd like to protect our Cloud Native Buildpack operations from rate limits and other service issues that may occur on public registries. While private mirrors are relatively easy to set up for k8s nodes workloads, it is difficult to configure Cloud Native Buildpacks to use them as images are requested while running inside `lifecycle` container processes. We'd like to be able to configure CNB to use a private mirror for the registries without affecting the resulting image.

Importantly, the presence of a mirror should be invisible outside of the platform. The resulting image should not contain any references to the mirror. This will allow the image to be used in any environment, regardless of whether the mirror is available. For example, the metadata set on the resulting image would reference the original registry URL, not the mirror URL. This will allow the resulting image to be used as if there was no mirror configured. This is important for future actions against the resulting image, such as `pack rebase` or `pack inspect`.

> - What use cases does it support?
> If a public registry was having service interruptions, an operator that had a previously configured mirror would be able to continue to build images without interruption.
> 
> If a public registry introduced lower rate limits, an operator that had a previously configured mirror would be able to continue to build images without interruption or fear of hitting the rate limit.
> 
> - What is the expected outcome?
> Operators concerned with the reliability of their builds that use public images will be able to configure private registry mirrors for the most popular registries.

Reviewer 2: (Fixes the spacing/indentation)

```suggestion
- What use cases does it support?

  If a public registry was having service interruptions, an operator that had a previously configured mirror would be able to continue to build images without interruption.

  If a public registry introduced lower rate limits, an operator that had a previously configured mirror would be able to continue to build images without interruption or fear of hitting the rate limit.

- What is the expected outcome?

  Operators concerned with the reliability of their builds that use public images will be able to configure private registry mirrors for the most popular registries.
```

```
docker.io=https://docker.mirror.example.com;quay.io=https://quay.mirror.example.com
```

When `lifecycle` requests an image during any phase (`analyze`, `restore`, `export`, `rebase`), it will first check the `CNB_REGISTRY_MIRRORS` environment variable. If the requested image's registry is configured in the `CNB_REGISTRY_MIRRORS` environment variable, it will use the mirror URL instead of the original registry URL.

If the private registry requires authentication, authentication to the registry will be handled by the existing `CNB_REGISTRY_AUTH` value. If the private registry does not require authentication, no additional configuration is required.

If registry mirrors are configured for specific images in configuration (e.g. `stack.toml` or `run.toml`), the `CNB_REGISTRY_MIRRORS` will be processed with each registry attempt. For example, the following `stack.toml` configures mirrors for `public/stack:run-image`:

```toml
[run-image]
 image = "public/stack:run-image"
 mirrors = ["quay.io/public/stack:run-image"]
```

If `CNB_REGISTRY_MIRRORS` has the value of:

```
docker.io=https://docker.mirror.example.com;quay.io=https://quay.mirror.example.com
```

> When `lifecycle` attempts to resolve the `public/stack:run-image` image, `lifecycle` will attempt to fetch the image from `docker.mirror.example.com/public/stack:run-image`. If the [Run Image Resolution](https://github.com/buildpacks/spec/blob/main/platform.md#run-image-resolution) resulted in `quay.io/public/stack:run-image` being chosen, `lifecycle` will attempt to fetch the image from `quay.mirror.example.com/public/stack:run-image` instead. This new processing happens AFTER the [Run Image Resolution](https://github.com/buildpacks/spec/blob/main/platform.md#run-image-resolution) has executed.  The run image selection will NOT take `CNB_REGISTRY_MIRRORS` into account, but the final image resolution will. Think of `CNB_REGISTRY_MIRRORS` as a just-in-time override of the final image resolution.

Reviewer: Just thinking out loud here, I'm not sure if this is relevant or not...

There exists today a `pack config run-image-mirrors` that will
* [transparently](https://github.com/buildpacks/pack/blob/3364a6b4b8b8aab03e90722f479a918f9dd3a8a9/pkg/client/build.go#L324) translate the run image reference

There is also a `pack config registry-mirrors` that will 
* [quietly](https://github.com/buildpacks/pack/blob/3364a6b4b8b8aab03e90722f479a918f9dd3a8a9/pkg/client/client.go#L218) translate the registry before fetching ANY image for ANY operation
* [transparently](https://github.com/buildpacks/pack/blob/3364a6b4b8b8aab03e90722f479a918f9dd3a8a9/pkg/client/build.go#L482) translate the registry for the run image reference AFTER `run-image-mirrors` have been applied

It's worth noting that while these translations update the `-run-image` provided to the lifecycle, they do NOT update stack.toml/run.toml i.e., these mirrors are "hidden" today in the sense that they are not persisted as metadata on the app image.

Reviewer: Now I remember - it would be helpful to describe how `CNB_REGISTRY_MIRRORS` env var for the lifecycle would be exposed in pack given the above stuff that exists today. I would really hope that we could avoid adding yet another toggle here.

Reviewer: @buildpacks/platform-maintainers @jjbustamante do you have any thoughts here?

Reviewer 3: I tried to search in past issues and pull request to understand the reason for `pack config mirrors` commands. I found the following:

The order of preference applied for for run image when `pack build` is executed:

1. --run-image flag
2. default run image mirror, as discussed in the --publish/daemon cases above
    2.1 when `publish`, default to using the run-image mirror in the registry you are publishing the image to  
    2.2 when `daemon`, defaults to using the run-image mirror in the same registry (colocated) as the builder 
3. mirrors set using `pack config run-image-mirrors` 
4. mirrors defined on the builder

Based on this, the new `CNB_REGISTRY_MIRRORS` seems to be like a new option to override point **4**, right? something like:

4. expose some way to set `CNB_REGISTRY_MIRRORS` and pass it through the lifecycle
5. mirrors defined on the builder

Why should we *not* do this?

Complexity. We'll have to teach `lifecycle` how to use the configured mirrors. This will add complexity to the codebase and will require additional testing.

Breaking expectations. If a platform operator were to fall behind or modify an image in the mirror, the resulting image would not match the image built by an end user against the public registry.

# Alternatives
[alternatives]: #alternatives

- What other designs have been considered?

Platforms could use oci layout to try and fetch images from a local cache before going to the public registry. This would require the platform to have a local cache of all images that may be requested during builds. This would be difficult to maintain and would require a lot of disk space.

- Why is this proposal the best?

This may not be the best, but that is why we are proposing it. We'd like to hear from the community about other options.

- What is the impact of not doing this?

Operators will have to continue to deal with the reliability of public registries.

# Prior Art
[prior-art]: #prior-art

Discuss prior art, both the good and bad.

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> - How will will teach kaniko-style extensions to use the mirrors?

Reviewer 2: ```suggestion
- How will we teach kaniko-style extensions to use the mirrors?
```
